chaos_genius/controllers/kpi_controller.py

```diff
@@ -58,45 +53,47 @@ def get_kpi_data_from_id(n: int) -> dict:
 def run_anomaly_for_kpi(
     kpi_id: int, end_date: Optional[date] = None, task_id: Optional[int] = None
 ):
+    """Runs anomaly detection for given kpi_id.
 
+    Blocking function (it does NOT spawn a celery task).
+    """
     logger.info(f"Starting Anomaly Detection for KPI ID: {kpi_id}.")
     kpi_info = get_kpi_data_from_id(kpi_id)
     logger.info(f"(KPI ID: {kpi_id}) Retrieved KPI information.")
 
     logger.info("(KPI ID: {kpi_id}) Selecting end date.")
 
-    if (
-        end_date is None
-        and kpi_info["scheduler_params"]["scheduler_frequency"] == "D"
-    ):
+    if end_date is None and kpi_info["scheduler_params"]["scheduler_frequency"] == "D":
         # by default we always calculate for n-days_offset_for_analytics
-        end_date = datetime.today().date() - timedelta(
+        true_end_date = datetime.today().date() - timedelta(
             days=(DAYS_OFFSET_FOR_ANALTYICS)
         )
         # Check if data is available or not then try for n-days_offset_for_analytics-1
-        if not _is_data_present_for_end_date(kpi_info, end_date):
-            end_date = end_date - timedelta(days=1)
+        if not _is_data_present_for_end_date(kpi_info, true_end_date):
+            true_end_date = true_end_date - timedelta(days=1)
             logger.info("(KPI ID: {kpi_id}) Decreasing end date by 1.")
 
     elif (
-        end_date is None
-        and kpi_info["scheduler_params"]["scheduler_frequency"] == "H"
+        end_date is None and kpi_info["scheduler_params"]["scheduler_frequency"] == "H"
     ):
-        end_date = datetime.today().date()
+        true_end_date = datetime.today().date()
+
+    else:
+        raise Exception(
+            "end_date isn't provided and scheduler_frequency is not recognized."
+        )
```

Comment:
I don't think we've handled the case where end_date is not None here. Here's my solution:
```python
if end_date is None:
    if scheduler_freq = "D":
        # create true_end_date for daily
    elif scheduler_freq = "H":
        # create true_end_date for hourly
    else:
        # raise error
else:
    true_end_date = end_date